docs/docs/examples.md

```diff
@@ -0,0 +1,317 @@
+
+!!! note inline end
+    Such constraints may seem artificial, but when bandwidth is low, a compact representation is key.
+
+A log message is an implicitly tagged ASN.1 structure with APPLICATION tag `1337` and sequence semantics.
+It contains the number of times an operation was run, and a timestamp, which can be either relative (in whole seconds since the last operation)
+or absolute (UTC Time).
+This relative/absolute flag uses the implicit APPLICATION tag `42` and the tuple of flag and time
+is encoded into an ANS.1 OCTET STRING. This allows for two possible encodings, as illustrated below:
+
+<table>
+<tr>
+<th>
+Absolute Time
+</th>
+<th>
+Relative Time
+</th>
+</tr>
+<tr>
+<td>
+
+```asn1
+Application 1337 (2 elem)
+  INTEGER 1
+  OCTET STRING (19 byte)
+    Application 42 (1 byte) 00
+    UTCTime 2024-09-30 18:11:59 UTC
+```
+
+</td>
+
+<td>
+
+```asn1
+Application 1337 (2 elem)
+  INTEGER 3
+  OCTET STRING (7 byte)
+    Application 42 (1 byte) FF
+    INTEGER 39
+```
+
+</td>
+</tr>
+</table>
+
+### Encoding
+
+We'll be assuming absolute time to keep things simple.
+Hence, the structure containing an absolute time can be created using the _Indispensable_ ASN.1 engine as follows:
+
+```kotlin
+val TAG_TIME_RELATIVE = 42uL withClass TagClass.APPLICATION
+
+Asn1.Sequence {
+    +Asn1.Int(1)
+    +OctetStringEncapsulating {
+        +(Bool(false) withImplicitTag TAG_TIME_RELATIVE)
+        +Asn1Time(Clock.System.now())
+    }
+} withImplicitTag (1337uL withClass TagClass.APPLICATION) 
+//                ↑ in reality this would be a constant ↑ 
+```
+
+The HEX-equivalent of this structure (which can be obtained by calling `.toDerHexString()`) is
+[7F8A391802010104135F2A0100170D3234303933303138313135395A](https://lapo.it/asn1js/#f4o5GAIBAQQTXyoBABcNMjQwOTMwMTgxMTU5Wg).
+
+### Parsing and Validating Tags
+
+Basic parsing is straight-forward: You have DER-encoded bytes, and feed them into `AsnElement.parse()`.
+In this example, you then examine the first child to get the number of times the operation was carried out,
+decode the first child of the OCTET STRING that follows to see if n UTC time or an int follows.
+
+Usually, though (and especially when using implicit tags), you really want to verify those tags too.
+Hence, parsing and properly validating is a bit more elaborate:
+
+```kotlin linenums="1"
+Asn1Element.parse(customSequence.derEncoded).asStructure().let { root -> 
+
+  //↓↓↓ In reality, this would be a global constant; the same as in the previous snippet ↓↓↓
+  val rootTag = Asn1Element.Tag(1337uL, tagClass = TagClass.APPLICATION, constructed = true)
+  root.assertTag(rootTag) //throws on tag mismatch
+
+  val numberOfOps = root.nextChild().asPrimitive().decodeToUInt()
+  root.nextChild().asEncapsulatingOctetString().let { timestamp ->
+    val isRelative = timestamp.nextChild().asPrimitive()
+      .decode(TAG_TIME_RELATIVE) { Boolean.decodeFromAsn1ContentBytes(it) }
```

Comment:
is there really no `.decodeBoolean` or similar? or even better, `.decode<Boolean>(tag)`?

Reply:
we have a generic decode for anything custom (i.e. implicitly tagged). `decodeBoolean` expects a boolean tag. as per documentation, decoding primitives is implemented as `<Primitive>.decodeFromAsn1ContentBytes(bytes)`

Reply:
then `.decodeBoolean(tag: Tag = TAG_BOOLEAN)`? this is a lot of characters for a very simple operation

Reply:
essentially, you want every  `.decodeXXX` function to have a tag override? sounds reasonable!

Reply:
done